client/daemon/rpcserver/rpcserver.go

```diff
@@ -295,3 +310,237 @@ func (s *server) Download(ctx context.Context,
 		}
 	}
 }
+
+func (s *server) StatTask(ctx context.Context, req *dfdaemongrpc.StatTaskRequest) (*base.GrpcDfResult, error) {
+	code := base.Code_Success
+	msg := "task found in cache"
+	taskID := idgen.TaskID(req.Cid, req.UrlMeta)
+	log := logger.With("function", "StatTask", "Cid", req.Cid, "Tag", req.UrlMeta.Tag, "taskID", taskID, "LocalOnly", req.LocalOnly)
+
+	log.Info("new stat task request")
+	if completed := s.isTaskCompleted(taskID); !completed {
+		// If only stat local cache and task doesn't exist, return not found
+		if req.LocalOnly {
+			msg = "task not found in local cache"
+			log.Info(msg)
+			return common.NewGrpcDfResult(base.Code_PeerTaskNotFound, msg), nil
+		}
+		res, err := s.peerTaskManager.StatPeerTask(ctx, taskID)
+		if err != nil {
+			msg = fmt.Sprintf("failed to StatPeerTask from peers: %v", err)
+			log.Error(msg)
+			return nil, errors.New(msg)
+		}
+		if res.Code == base.Code_PeerTaskNotFound {
+			code = res.Code
+			msg = res.Message
+			log.Info(msg)
+		}
+	}
+	return common.NewGrpcDfResult(code, msg), nil
+}
+
+func (s *server) ImportTask(ctx context.Context, req *dfdaemongrpc.ImportTaskRequest) (*base.GrpcDfResult, error) {
+	peerID := idgen.PeerID(s.peerHost.Ip)
+	taskID := idgen.TaskID(req.Cid, req.UrlMeta)
+	log := logger.With("function", "ImportTask", "Cid", req.Cid, "Tag", req.UrlMeta.Tag, "taskID", taskID, "file", req.Path)
+
+	log.Info("new import task request")
+	ptm := storage.PeerTaskMetadata{
+		PeerID: peerID,
+		TaskID: taskID,
+	}
+	announceFunc := func() {
+		// TODO: retry announce on error
+		start := time.Now()
+		err := s.peerTaskManager.AnnouncePeerTask(context.Background(), ptm, req.Cid, req.UrlMeta)
+		if err != nil {
+			log.Warnf("Failed to announce task to scheduler: %s", err)
+		} else {
+			log.Infof("Announce task (peerID %s) to scheduler in %.6f seconds", ptm.PeerID, time.Since(start).Seconds())
+		}
+	}
+
+	// 0. Task exists in local storage
+	if task := s.storageManager.FindCompletedTask(taskID); task != nil {
+		msg := fmt.Sprintf("import file skipped, task already exists with peerID %s", task.PeerID)
+		log.Info(msg)
+
+		// Announce to scheduler as well, but in background
+		ptm.PeerID = task.PeerID
+		go announceFunc()
+		return common.NewGrpcDfResult(base.Code_Success, msg), nil
+	}
+
+	// 1. Register to storageManager
+	// TODO: compute and check hash digest if digest exists in ImportTaskRequest
+	if _, err := s.storageManager.RegisterTask(ctx, &storage.RegisterTaskRequest{
+		PeerTaskMetadata: storage.PeerTaskMetadata{
+			PeerID: peerID,
+			TaskID: taskID,
+		},
+	}); err != nil {
+		msg := fmt.Sprintf("register task to storage manager failed: %v", err)
+		log.Error(msg)
+		return nil, errors.New(msg)
+	}
+
+	// 2. Import task file
+	pieceManager := s.peerTaskManager.GetPieceManager()
+	if err := pieceManager.ImportFile(ctx, ptm, s.storageManager, req); err != nil {
```

Comment:
s.storageManager.RegisterTask returns a TaskStorageDriver, it can be used here:
```
tsd, err := s.storageManager.RegisterTask(...)
pieceManager.ImportFile(ctx, ptm, tds req)
```

Reply:
Fixed.